Notify slack when release phase fails

mstruve opened the pull request:
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
Currently the only way we are notified of release phase failures is via an email from Heroku to the Heroku admins. Unfortunately, that's not helpful to anyone else who is trying to deploy their code. This attempts to hook up a command in our release phase script to catch any errors that occur and notify slack about them for everyone to see. 

I am FAR from an expert in bash and actually [found this script on no other than StackOverflow](https://stackoverflow.com/a/41852435/1438582). There is probably a slicker better way to do this so please let me know what that is! 

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/projects/5#card-33642543

## Added tests?
- [x] yes

## Added to documentation?
- [x] no documentation needed

![alt_text](https://media3.giphy.com/media/l4lRmZ9JXumY3q2zu/giphy.gif)

citizen428 commented:
@mstruve Here's a solution using Bash's built-in error handling. 

For this to work, we need a few things.

1. Set the `-e` option, so Bash will exit immediately on error.
2. Set up a [trap](https://www.linuxjournal.com/content/bash-trap-command), but only in the case of errors, not regular exits (the `ERR` signal)
3. Use Bash's `caller` built-in and `$BASH_COMMAND` for reporting.

Applying these changes, the script looks like this:

```sh
#!/bin/bash

notify () {
  FAILED_COMMAND="$(caller): ${BASH_COMMAND}" \
    bundle exec rails runner "ReleasePhaseNotifier.ping_slack"
}

trap notify ERR

# enable echo mode (-x) and exit on error (-e)
set -ex

# abort release if deploy status equals "blocked"
[[ $DEPLOY_STATUS = "blocked" ]] && echo "Deploy blocked" && exit 1

# runs migration for Postgres, setups/updates Elasticsearch
# and boots the app to check there are no errors
STATEMENT_TIMEOUT=180000 bundle exec rails db:migrate
bundle exec rake search:setup
bundle exec rake data_updates:enqueue_data_update_worker
bundle exec rails runner "puts 'app load success'"
```

I then added an invalid command `foo` before the migration to demonstrate how this works:

```sh
+ [[ '' = \b\l\o\c\k\e\d ]]
+ foo
release-tasks.sh: line 18: foo: command not found
++ notify
+++ caller
++ FAILED_COMMAND='18 release-tasks.sh: foo'
++ bundle exec rails runner ReleasePhaseNotifier.ping_slack
```

As you can see `FAILED_COMMAND contains the line number (18), the filename (`release-tasks.sh`and the command that failed (`foo`). If you don't want the first bit in the output just remove the `caller` part (or replace it with `$LINENO`):

```sh
notify () {
  FAILED_COMMAND="${LINENO}: ${BASH_COMMAND}" \
    bundle exec rails runner "ReleasePhaseNotifier.ping_slack"
}
```

Result

```
++ FAILED_COMMAND=4: foo
```

mstruve commented:
THANK YOU @citizen428 for the Bash tutorial! Implemented your changes and then ran a few tests and it looks like it works perfectly. Here is an example of an early failure. I changed ping to just use puts instead of talking to slack
```
mollystruve$ bash release-tasks.sh
+ [[ '' = \b\l\o\c\k\e\d ]]
+ STATEMENT_TIMEOUT=180000
+ bundle exec rails db:migrate
Dumping database schema with fix-db-schema-conflicts gem
+ bundle exec rake search:setup
rake aborted!
BROKEN!!!!
/Users/mollystruve/dev.to/lib/tasks/search.rake:4:in `block (2 levels) in <main>'
/Users/mollystruve/.rvm/gems/ruby-2.6.5/gems/honeycomb-beeline-1.3.0/lib/honeycomb/integrations/rake.rb:21:in `block in execute'
/Users/mollystruve/.rvm/gems/ruby-2.6.5/gems/honeycomb-beeline-1.3.0/lib/honeycomb/client.rb:54:in `start_span'
/Users/mollystruve/.rvm/gems/ruby-2.6.5/gems/honeycomb-beeline-1.3.0/lib/honeycomb/integrations/rake.rb:16:in `execute'
/Users/mollystruve/.rvm/gems/ruby-2.6.5/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
/Users/mollystruve/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
/Users/mollystruve/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
Tasks: TOP => search:setup
(See full trace by running task with --trace)
++ notify
+++ caller
++ FAILED_COMMAND='19 release-tasks.sh: bundle exec rake search:setup'
++ bundle exec rails runner ReleasePhaseNotifier.ping_slack
Release Phase Failed: 19 release-tasks.sh: bundle exec rake search:setup
```

rhymes commented on app/lib/release_phase_notifier.rb:
```diff
@@ -0,0 +1,11 @@
+      ApplicationConfig["SLACK_WEBHOOK_URL"],
+      channel: ApplicationConfig["SLACK_DEPLOY_CHANNEL"],
+      username: "Heroku",
+    )
```
I get `undefined method `request_uri' for #<URI::Generic >` when I run this, because by default `SLACK_WEBHOOK_URL` is empty, I think we should bail from this method if they aren't both defined

mstruve replied:
Took it a step further and added error handling for Slack API errors as well so if Slack goes down we can still deploy. I would be persuaded to catch any `StandardError` if you all think that would be better.

rhymes commented on app/lib/release_phase_notifier.rb:
```diff
@@ -0,0 +1,15 @@
+
+    client.ping("Release Phase Failed: #{ENV['FAILED_COMMAND']}")
+  rescue Slack::Notifier::APIError => e
+    Rails.logger.info("Slack API error occured: #{e.message}")
```
`Rails.logger.error` ?

mstruve replied:
Log level is debug in production and all we need is to see it in the little log file that the script outputs on Heroku so I figured this would do

citizen428 commented on release-tasks.sh:
```diff
@@ -1,14 +1,21 @@
+trap notify ERR
+
+# enable echo mode (-x) and exit on error (-e)
+set -ex
```
@mstruve Let's please also add the `-E` option here. This will ensure that `ERR` traps get inherited by functions, command substitutions, and subshell environments. We don't need any of that in the current form, but it will make the script more robust.

mstruve replied:
Done! Thanks!
